core/image.h

```diff
@@ -477,4 +477,6 @@ template <class ImageType> typename enable_if_image_type<ImageType, void>::type
     WARN(std::string("error invoking viewer: ") + strerror(errno));
 }
 
+extern template MR::Image<bool>::~Image();
+extern template MR::Image<float>::~Image();
```

Comment:
These are probably the two most frequent, but would it make more sense to be exhaustive here?

I'm also thinking about the prospect of external projects compiling against core. Previously, straight use of templates would allow such projects to instantiate whatever alternative template usages they wished. We've not yet fleshed out exactly how external projects will work under `cmake`, but should be wary to not break such. With the syntax I've historically used myself for explicit instantiation of templates, that list of instantiations would need to be exhaustive; I'm inferring from this syntax that a similar problem *doesn't* occur here? Ie. The `extern template` says that that specific instantiation appears in the external file, but any other template instantiation can occur within any compilation job?

Reply:
> These are probably the two most frequent, but would it make more sense to be exhaustive here?

Yes, I guess that would be sensible although I'm not sure whether it would provide any build time benefits. Perhaps it makes sense to add one for `double` or are there others that you were thinking about?

> any other template instantiation can occur within any compilation job?

Yes, that's correct.

Reply:
- The fixel format uses `Image<uint32_t>` for the index image; `label`-related commands will also use `Image<uint32_t>`.
- `Image<double>` is probably used somewhere but nothing's coming to mind.
  (Arguably should be used in some locations where scratch images are used for intermediate results; and ideally in my mind, for many workhorse commands, if any input data are double precision then the main processing task should be templated to operate in double precision)
- I see `Image<uint16_t>` used in `dwidenoise`.
- `mrcalc` uses `Image<cfloat>`; I see `mrfilter` using `Image<cdouble>`.

Possibly what would make sense would be to instantiate all possibilities that are reflected in `MR::DataType`. That would intrinsically communicate the *intended* scope of template types (which has caused problems in the couple of instances where it's been violated), and would hopefully mean that if any external project were to make use of the shared library than any reasonably expected image type should already be there.

Reply:
I don't think that `MR::DataType` would be a suitable place to instantiate these templates. Note that we are not instantiating the types `MR::Image<float>` or `MR::Image<bool>`, but the *destructors* of those classes (odd I know, but `ClangBuildAnalyzer` showed that they took a long time to instantiate). I think it would look very weird if the destructor of a class appeared to be in a completely unrelated source file from its header.

Reply:
Not suggesting that the relevant code should appear in `MR::DataType`, rather that the set of templates instantiated for `MR::Image<>` should echo the data type possibilities enumerated in `MR::DataType` (omitting endianness). Both are then exhaustive of the possible data types that could be used to store image data.

So:
- `bool`
- `int8_t`
- `uint8_t`
- `int16_t`
- `uint16_t`
- `int32_t`
- `uint32_t`
- `int64_t`
- `uint64_t`
- `half`
- `float`
- `double`
- `cfloat`
- `cdouble`